refnx/analysis/curvefitter.py

```diff
@@ -535,7 +535,7 @@ def _callback_wrapper(state, h=None):
         # sets parameter value and stderr
         return process_chain(self.objective, self.chain)
 
-    def fit(self, method='L-BFGS-B', **kws):
+    def fit(self, method='L-BFGS-B', prior=False, **kws):
```

Comment:
I'd prefer to use a *slightly* different approach. I'm not much good with naming, but how about we use a `target='nll'` keyword. The options for the `target` keyword would be `{'nll', 'nlpost'}`.
Default would be to do the maximum log-likelihood, but the other option would be maximise the log-posterior (often referred to as MAP).
I think that `nlpost` will fail if gradient based minimisers stray into an area for which the prior is non-finite (`-np.inf`), otherwise they should be fine. Differential evolution should be able to cope with non-finite priors so long as they're not ubiquitous.

I think the biggest challenge is the use of a `PDF` object to specify the prior/bounds for a given parameter. Let's say the `PDF` is Gaussian. It's easy to calculate the probabilities for that. The issue comes when an optimisation method (e.g. differential evolution) mandates lower and upper bounds for the parameter. The fitter [looks to each parameter](https://github.com/refnx/refnx/blob/master/refnx/analysis/curvefitter.py#L846) to see if its bounds attribute is an `Interval`. If it's not an `Interval`, which is won't be if you use a `PDF`, then the fitting process won't work because there will be no lower and upper bounds. I'm not sure how to get around this challenge, it'll require some thought.

TLDR: Not sure if it's possible to use anything other than a non-uniform distribution with a method that works on box (lower, upper) bounds.

Reply:
It's only going to work on the scalar minimisers, not least-squares.